File: Heat/Controllers/PlantsController.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Entity
Imports System.Linq
Imports System.Net
Imports System.Web
Imports System.Web.Mvc
Imports Heat
Imports Heat.Repositories
Imports Heat.Models
Imports Heat.ViewModels.Plants
Imports DataTables.AspNet.Core
Imports Heat.Manager
Imports System.IO

Namespace Controllers
    <Authorize>
    Public Class PlantsController
        Inherits System.Web.Mvc.Controller

        Private _db As IHeatDBContext
        Private _mb As PlantModelViewBuilder
        Private _pm As PlantManager


        Sub New(dbcontext As IHeatDBContext)
            _db = dbcontext
            _mb = New PlantModelViewBuilder(_db)
            _pm = New PlantManager(_db)
        End Sub


        ' GET: Plants
        Function Index() As ActionResult
            Try
                Dim model As List(Of IndexPlantViewModel)

                model = _mb.GetIndexPlantViewModel

                Return View(model)
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpGet>
        Function Details(ByVal id As Integer?) As ActionResult
            Try
                If IsNothing(id) Then
                    Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                End If
                'Dim plant As Plant = _db.Plants.Find(id)
                If Not _db.Plants.Any(Function(p) p.ID = id) Then
                    Return HttpNotFound()
                End If
                Dim model As DetailsPlantViewModel
                model = _mb.GetDetailsPlantViewModel(id)
                Return View(model)
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpGet>
        Function Create() As ActionResult
            Try

                Return View()
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try
        End Function


        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Create(newPlant As CreatePlantViewModel) As ActionResult
            Try
                If ModelState.IsValid Then
                    Dim p As Plant

                    p = AutoMapper.Mapper.Map(Of Plant)(newPlant)

                    _db.Plants.Add(p)
                    _db.SaveChanges()
                    'dopo aver salvato i dati dell'ubicazione passo ai dati del contatto
                    Return RedirectToAction("AddContact", New With {.plantID = p.ID})
                Else
                    'il modello non è valido
                    Return View(newPlant)
                End If

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpGet>
        Function AddContact(ID As Integer) As ActionResult
            Try
                If IsNothing(ID) Then
                    Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                End If
                'Dim plant As Plant = _db.Plants.Find(plantID)
                If Not _db.Plants.Any(Function(p) p.ID = ID) Then
                    Return HttpNotFound()
                End If

                Dim model As AddContactPlantViewModel

                model = _mb.GetAddContactPlantViewModel(ID)
                Return View(model)

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpPost>
        <ValidateAntiForgeryToken>
        Function AddContact(newContact As AddContactPlantViewModel) As ActionResult
            Try
                If ModelState.IsValid Then
                    Dim c As Contact
                    Dim p As Plant

                    c = AutoMapper.Mapper.Map(Of Contact)(newContact)
                    p = _db.Plants.Find(newContact.PlantID)

                    If Not IsNothing(p) Then
                        p.Contacts.Add(c)
                        _db.Contacts.Add(c)
                        _db.SaveChanges()
                        'dopo aver salvato i dati del contatto passo ai dati termici
                        Return RedirectToAction("AddThermInfo", New With {.plantID = p.ID})
                    Else
                        ViewBag.message = "Impossibile aggiungere il contatto. Sembra che l'impianto con id(" & newContact.PlantID & ") non esista"
                        Return View("error")
                    End If

                Else
                    'ripasso al modelBuilder per ricostruire la selectlist
                    Return View(_mb.GetAddContactPlantViewModel(newContact.PlantID))
                End If

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try
        End Function

        <HttpGet>
        Function AddAnotherContact(ID As Integer) As ActionResult
            Try
                If IsNothing(ID) Then
                    Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                End If

                If Not _db.Plants.Any(Function(p) p.ID = ID) Then
                    Return HttpNotFound()
                End If

                Dim model As AddContactPlantViewModel

                model = _mb.GetAddContactPlantViewModel(ID)
                Return View(model)

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpPost>
        <ValidateAntiForgeryToken>
        Function AddAnotherContact(newContact As AddContactPlantViewModel) As ActionResult
            Try
                If ModelState.IsValid Then
                    Dim c As Contact
                    Dim p As Plant

                    c = AutoMapper.Mapper.Map(Of Contact)(newContact)
                    p = _db.Plants.Find(newContact.PlantID)

                    If Not IsNothing(p) Then
                        p.Contacts.Add(c)
                        _db.Contacts.Add(c)
                        _db.SaveChanges()
                        'dopo aver salvato i dati del contatto torno ai dettagli dell'impianto
                        Return RedirectToAction("details", New With {.ID = p.ID})
                    Else
                        ViewBag.message = "Impossibile aggiungere il contatto. Sembra che l'impianto con id(" & newContact.PlantID & ") non esista"
                        Return View("error")
                    End If

                Else
                    'ripasso al modelBuilder per ricostruire la selectlist
                    Return View(_mb.GetAddContactPlantViewModel(newContact.PlantID))
                End If

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try
        End Function

        <HttpGet>
        Function AddThermInfo(plantId As Integer) As ActionResult
            Try
                If IsNothing(plantId) Then
                    Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                End If
                Dim plant As Plant = _db.Plants.Find(plantId)
                If IsNothing(plant) Then
                    Return HttpNotFound()
                End If
                Dim model As AddThermInfoPlantViewModel

                model = _mb.GetAddThermInfoPlantViewModel(plantId)
                Return View(model)

            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpPost>
        Function AddThermInfo(newThermInfo As AddThermInfoPlantViewModel) As ActionResult
            Try
                If ModelState.IsValid Then
                    Dim ActualPB As PlantBuilding
                    Dim actualPlant As Plant

                    actualPlant = _db.Plants.Find(newThermInfo.PlantID)
                    ActualPB = actualPlant.BuildingAddress

                    ActualPB = AutoMapper.Mapper.Map(Of AddThermInfoPlantViewModel, PlantBuilding)(newThermInfo, ActualPB)
                    actualPlant.PlantDistinctCode = newThermInfo.PlantDistinctCode

                    _db.SaveChanges()

                    Return RedirectToAction("index")
                Else
                    'ripasso al modelBuilder per ricreare le selectlist
                    Return View(_mb.GetAddThermInfoPlantViewModel(newThermInfo.PlantID))
                End If
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try
        End Function

        ' GET: Plants/Edit/5
        Function Edit(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim plant As Plant = _db.Plants.Find(id)
            If IsNothing(plant) Then
                Return HttpNotFound()
            End If
            Return View(plant)
        End Function

        ' POST: Plants/Edit/5
        'To protect from overposting attacks, please enable the specific properties you want to bind to, for 
        'more details see http://go.microsoft.com/fwlink/?LinkId=317598.
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Edit(<Bind(Include:="ID,Code,Name,Address,StreetNumber,City,PostalCode,Area,Zone,PlantTelephone1,PlantTelephone2,PlantTelephone3,PlantDistictCode,Fuel")> ByVal plant As Plant) As ActionResult
            If ModelState.IsValid Then
                '_db.Entry(plant).State = EntityState.Modified
                _db.SaveChanges()
                Return RedirectToAction("Index")
            End If
            Return View(plant)
        End Function

        ' GET: Plants/Delete/5
        Function Delete(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim plant As Plant = _db.Plants.Find(id)
            If IsNothing(plant) Then
                Return HttpNotFound()
            End If
            Return View(plant)
        End Function

        ' POST: Plants/Delete/5
        <HttpPost()>
        <ActionName("Delete")>
        <ValidateAntiForgeryToken()>
        Function DeleteConfirmed(ByVal id As Integer) As ActionResult
            Dim plant As Plant = _db.Plants.Find(id)
            _db.Plants.Remove(plant)
            _db.SaveChanges()
            Return RedirectToAction("Index")
        End Function

        <HttpPost>
        Function Import(uploadFilePlant As HttpPostedFileBase) As ActionResult
            If Not IsNothing(uploadFilePlant) AndAlso uploadFilePlant.ContentLength > 0 Then
                Dim fileExt As String
                fileExt = System.IO.Path.GetExtension(uploadFilePlant.FileName).ToLower
                If fileExt = ".txt" Then
                    Dim ih As New ImportHelper(_db)
                    Dim b(uploadFilePlant.ContentLength) As Byte
                    uploadFilePlant.InputStream.Read(b, 0, uploadFilePlant.ContentLength)
                    If ih.Plant(System.Text.Encoding.ASCII.GetString(b)) Then
                        Return RedirectToAction("index")
                    Else
                        ViewBag.error = "Errore durante l'importazione del file"
                        Return View()
                    End If
                Else
                    ViewBag.error = "Sono ammmessi solo file .txt"
                    Return View()
                End If

            Else
                Return View()
            End If
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="request"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <HttpGet>
        Public Function PagedPlants(request As IDataTablesRequest) As ActionResult
            If ModelState.IsValid Then
                Return _pm.GetPagedPlants(request)
            Else
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                'Return Json(New With {.success = False, .responseText = "La richiesta non è valida!"}, JsonRequestBehavior.AllowGet)
            End If
        End Function

        <HttpGet>
        Public Function AddMediumToPlant(plantID As Integer) As ActionResult
            If IsNothing(plantID) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If

            If Not _db.Plants.Any(Function(p) p.ID = plantID) Then
                Return HttpNotFound()
            End If
            Try
                Dim model As New AddMediumPlantViewModel
                model.PlantId = plantID
                Return View(model)
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        <HttpPost>
        <ValidateAntiForgeryToken>
        Public Function AddMediumToPlant(newMedium As AddMediumPlantViewModel) As ActionResult
            If IsNothing(newMedium.UploadFile) OrElse newMedium.UploadFile.ContentLength <= 0 Then
                ModelState.AddModelError("UploadFile", "Il file è richiesto!")
            End If
            Try
                If ModelState.IsValid Then
                    Dim medium As New Medium
                    Dim uploadDir As String
                    Dim uploadFileName As String
                    Dim uploadPath As String
                    Dim uploadURL As String

                    uploadDir = ConfigurationManager.AppSettings("MediaPlantFolder")
                    uploadFileName = Guid.NewGuid().ToString & Path.GetExtension(newMedium.UploadFile.FileName)

                    uploadPath = Path.Combine(Server.MapPath(uploadDir), uploadFileName)
                    uploadURL = Path.Combine(uploadDir, uploadFileName)

                    newMedium.UploadFile.SaveAs(uploadPath)

                    medium.Description = newMedium.Description
                    medium.Tags = newMedium.Tags
                    medium.OriginalFilename = newMedium.UploadFile.FileName
                    medium.UploadFilename = uploadFileName
                    medium.ContentType = newMedium.UploadFile.ContentType
                    medium.Lenght = newMedium.UploadFile.ContentLength

                    Dim p As Plant = _db.Plants.Find(newMedium.PlantId)
                    p.Media.Add(medium)
                    _db.Media.Add(medium)
                    _db.SaveChanges()
                    Return RedirectToAction("index")
                Else
                    Return View(newMedium)
                End If
            Catch ex As Exception
                ViewBag.message = ex.ToString
                Return View("error")
            End Try

        End Function

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If (disposing) Then
                _db.Dispose()
            End If
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace

File: Heat.ConvertedToC#/Controllers/PlantsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Configuration;
using System.Web;
using System.Web.Mvc;
using System.Net;
using Heat.Models;
using Heat.ViewModels.Plants;
using DataTables.AspNet.Core;
using Heat.Manager;
using System.IO;
using System.Text;

namespace Heat.Controllers
{
    [Authorize()]
	public class PlantsController : System.Web.Mvc.Controller
	{

		private IHeatDBContext _db;
		private PlantModelViewBuilder _mb;

		private PlantManager _pm;

		public PlantsController(IHeatDBContext dbcontext)
		{
			_db = dbcontext;
			_mb = new PlantModelViewBuilder(_db);
			_pm = new PlantManager(_db);
		}


		// GET: Plants
		public ActionResult Index()
		{
			try {
				List<IndexPlantViewModel> model = null;

				model = _mb.GetIndexPlantViewModel();

				return View(model);
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpGet()]
		public ActionResult Details(int  id)
		{
			try {
				if ((id == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}
				if (!_db.Plants.Any(p => p.ID == id)) {
					return HttpNotFound();
				}
				DetailsPlantViewModel model = null;
				model = _mb.GetDetailsPlantViewModel(id);
				return View(model);
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpGet()]
		public ActionResult Create()
		{

			try {
				return View();
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}
		}

		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult Create(CreatePlantViewModel newPlant)
		{
			try {
				if (ModelState.IsValid) {
					Plant p = null;

					p = AutoMapper.Mapper.Map<Plant>(newPlant);

					_db.Plants.Add(p);
					_db.SaveChanges();
					//dopo aver salvato i dati dell'ubicazione passo ai dati del contatto
					return RedirectToAction("AddContact", new { plantID = p.ID });
				} else {
					//il modello non è valido
					return View(newPlant);
				}

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpGet()]
		public ActionResult AddContact(int ID)
		{
			try {
				if ((ID == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}
				//Dim plant As Plant = _db.Plants.Find(plantID)
				if (!_db.Plants.Any(p => p.ID == ID)) {
					return HttpNotFound();
				}

				AddContactPlantViewModel model = null;

				model = _mb.GetAddContactPlantViewModel(ID);
				return View(model);

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult AddContact(AddContactPlantViewModel newContact)
		{
			try {
				if (ModelState.IsValid) {
					Contact c = null;
					Plant p = null;

					c = AutoMapper.Mapper.Map<Contact>(newContact);
					p = _db.Plants.Find(newContact.PlantID);

					if ((p != null)) {
						p.Contacts.Add(c);
						_db.Contacts.Add(c);
						_db.SaveChanges();
						//dopo aver salvato i dati del contatto passo ai dati termici
						return RedirectToAction("AddThermInfo", new { plantID = p.ID });
					} else {
						ViewBag.message = "Impossibile aggiungere il contatto. Sembra che l'impianto con id(" + newContact.PlantID + ") non esista";
						return View("error");
					}

				} else {
					//ripasso al modelBuilder per ricostruire la selectlist
					return View(_mb.GetAddContactPlantViewModel(newContact.PlantID));
				}

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}
		}

		[HttpGet()]
		public ActionResult AddAnotherContact(int ID)
		{
			try {
				if ((ID == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}

				if (!_db.Plants.Any(p => p.ID == ID)) {
					return HttpNotFound();
				}

				AddContactPlantViewModel model = null;

				model = _mb.GetAddContactPlantViewModel(ID);
				return View(model);

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult AddAnotherContact(AddContactPlantViewModel newContact)
		{
			try {
				if (ModelState.IsValid) {
					Contact c = null;
					Plant p = null;

					c = AutoMapper.Mapper.Map<Contact>(newContact);
					p = _db.Plants.Find(newContact.PlantID);

					if ((p != null)) {
						p.Contacts.Add(c);
						_db.Contacts.Add(c);
						_db.SaveChanges();
						//dopo aver salvato i dati del contatto torno ai dettagli dell'impianto
						return RedirectToAction("details", new { ID = p.ID });
					} else {
						ViewBag.message = "Impossibile aggiungere il contatto. Sembra che l'impianto con id(" + newContact.PlantID + ") non esista";
						return View("error");
					}

				} else {
					//ripasso al modelBuilder per ricostruire la selectlist
					return View(_mb.GetAddContactPlantViewModel(newContact.PlantID));
				}

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}
		}

		[HttpGet()]
		public ActionResult AddThermInfo(int plantId)
		{
			try {
				if ((plantId == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}
				Plant plant = _db.Plants.Find(plantId);
				if ((plant == null)) {
					return HttpNotFound();
				}
				AddThermInfoPlantViewModel model = null;

				model = _mb.GetAddThermInfoPlantViewModel(plantId);
				return View(model);

			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpPost()]
		public ActionResult AddThermInfo(AddThermInfoPlantViewModel newThermInfo)
		{
			try {
				if (ModelState.IsValid) {
					PlantBuilding ActualPB = null;
					Plant actualPlant = null;

					actualPlant = _db.Plants.Find(newThermInfo.PlantID);
					ActualPB = actualPlant.BuildingAddress;

					ActualPB = AutoMapper.Mapper.Map<AddThermInfoPlantViewModel, PlantBuilding>(newThermInfo, ActualPB);
					actualPlant.PlantDistinctCode = newThermInfo.PlantDistinctCode;

					_db.SaveChanges();

					return RedirectToAction("index");
				} else {
					//ripasso al modelBuilder per ricreare le selectlist
					return View(_mb.GetAddThermInfoPlantViewModel(newThermInfo.PlantID));
				}
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}
		}

		// GET: Plants/Edit/5
		public ActionResult Edit(int? id)
		{
			if ((id == null)) {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
			}
			Plant plant = _db.Plants.Find(id);
			if ((plant == null)) {
				return HttpNotFound();
			}
			return View(plant);
		}

		// POST: Plants/Edit/5
		//To protect from overposting attacks, please enable the specific properties you want to bind to, for 
		//more details see http://go.microsoft.com/fwlink/?LinkId=317598.
		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult Edit(		[Bind(Include = "ID,Code,Name,Address,StreetNumber,City,PostalCode,Area,Zone,PlantTelephone1,PlantTelephone2,PlantTelephone3,PlantDistictCode,Fuel")]
Plant plant)
		{
			if (ModelState.IsValid) {
				//_db.Entry(plant).State = EntityState.Modified
				_db.SaveChanges();
				return RedirectToAction("Index");
			}
			return View(plant);
		}

		/// <summary>
		/// Prepara la pagina con i link alle operazioni che è possibile svolgere sull'impianto.
		/// </summary>
		/// <param Name="id"></param>
		/// <returns></returns>
		[HttpGet()]
		public ActionResult Manage(int id)
		{
			try {
				if ((id == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}
				//Dim customer As Plant = _db.Plants.Find(id)
				if (!_db.Plants.Any(x => x.ID == id)) {
					return HttpNotFound();
				}
				ManagePlantViewModel m = null;
				m = _mb.GetManagePlantViewModel(id);
				return View(m);
			} catch (Exception ex) {
				ViewBag.message = ex.Message;
				return View("error");
			}
		}

		// GET: Plants/Delete/5
		public ActionResult Delete(int? id)
		{
			if ((id == null)) {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
			}
			Plant plant = _db.Plants.Find(id);
			if ((plant == null)) {
				return HttpNotFound();
			}
			return View(plant);
		}

		// POST: Plants/Delete/5
		[HttpPost()]
		[ActionName("Delete")]
		[ValidateAntiForgeryToken()]
		public ActionResult DeleteConfirmed(int id)
		{
			Plant plant = _db.Plants.Find(id);
			_db.Plants.Remove(plant);
			_db.SaveChanges();
			return RedirectToAction("Index");
		}

		[HttpPost()]
		public ActionResult Import(HttpPostedFileBase uploadFilePlant, HttpPostedFileBase uploadFileThermalUnit)
		{
			if (uploadFilePlant != null && uploadFilePlant.ContentLength > 0 && uploadFileThermalUnit != null && uploadFileThermalUnit.ContentLength>0) {
				string plantFileExt;
                string thermUnitFileExt;

				plantFileExt = Path.GetExtension(uploadFilePlant.FileName).ToLower();
                thermUnitFileExt = Path.GetExtension(uploadFileThermalUnit.FileName).ToLower();

                if (plantFileExt == ".txt" && thermUnitFileExt == ".txt") {
					ImportHelper ih = new ImportHelper(_db);
					byte[] plantsFileBytes = new byte[uploadFilePlant.ContentLength + 1];
                    byte[] thermalUnitsFileBytes = new byte[uploadFileThermalUnit.ContentLength + 1];

					uploadFilePlant.InputStream.Read(plantsFileBytes, 0, uploadFilePlant.ContentLength);
                    uploadFileThermalUnit.InputStream.Read(thermalUnitsFileBytes, 0, uploadFileThermalUnit.ContentLength);

                    string plantFileText = Encoding.ASCII.GetString(plantsFileBytes);
                    string thermalUnitFileText = Encoding.ASCII.GetString(thermalUnitsFileBytes);

					if (ih.Plant(plantFileText, thermalUnitFileText )) {
						return RedirectToAction("index");
					} else {
						ViewBag.error = "Errore durante l'importazione del file";
						return View();
					}
				} else {
					ViewBag.error = "Sono ammmessi solo file .txt";
					return View();
				}

			} else {
				return View();
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param Name="request"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		[HttpGet()]
		public ActionResult PagedPlants(IDataTablesRequest request)
		{
			if (ModelState.IsValid) {
				return _pm.GetPagedPlants(request);
			} else {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				//Return Json(New With {.success = False, .responseText = "La richiesta non è valida!"}, JsonRequestBehavior.AllowGet)
			}
		}

		/// <summary>
		/// Aggiunge un file all'impianto, come allegato.
		/// E' possibile allegare qualunque tipo di file.
		/// </summary>
		/// <param Name="ID">ID dell'impianto</param>
		/// <returns></returns>
		/// <remarks></remarks>
		[HttpGet()]
		public ActionResult AddMedium(int ID)
		{
			//http://cpratt.co/file-uploads-in-asp-net-mvc-with-view-models/
			if ((ID == null)) {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
			}

			if (!_db.Plants.Any(p => p.ID == ID)) {
				return HttpNotFound();
			}
			try {
				AddMediumPlantViewModel model = new AddMediumPlantViewModel();
				model.PlantId = ID;
				return View(model);
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult AddMedium(AddMediumPlantViewModel newMedium)
		{
			if ((newMedium.UploadFile == null) || newMedium.UploadFile.ContentLength <= 0) {
				ModelState.AddModelError("UploadFile", "Il file è richiesto!");
			}

			string[] validImageTypes = new string[] {
				"image/gif",
				"image/jpeg",
				"image/pjpeg",
				"image/png"
			};
			if (!validImageTypes.Contains(newMedium.UploadFile.ContentType)) {
				ModelState.AddModelError("UploadFile", "Questo tipo di file non è ancora supportato!");
			}


			try {
				if (ModelState.IsValid) {
					Medium medium = new Medium();
					string uploadDir = null;
					string uploadFileName = null;
					string uploadPath = null;
					string uploadURL = null;

					uploadDir = ConfigurationManager.AppSettings["MediaPlantFolder"];
					uploadFileName = Guid.NewGuid().ToString() + Path.GetExtension(newMedium.UploadFile.FileName);

					uploadPath = Path.Combine(Server.MapPath(uploadDir), uploadFileName);
					uploadURL = Path.Combine(uploadDir, uploadFileName);

					newMedium.UploadFile.SaveAs(uploadPath);

					medium.Description = newMedium.Description;
					medium.Tags = newMedium.Tags;
					medium.OriginalFileName = newMedium.UploadFile.FileName;
					medium.UploadFileName = uploadFileName;
					medium.ContentType = newMedium.UploadFile.ContentType;
					medium.Lenght = newMedium.UploadFile.ContentLength;

					Plant p = _db.Plants.Find(newMedium.PlantId);
					p.Media.Add(medium);

					_db.Media.Add(medium);
					_db.SaveChanges();
					return RedirectToAction("details", new { id = p.ID });
				} else {
					return View(newMedium);
				}
			} catch (Exception ex) {
				ViewBag.message = ex.ToString();
				return View("error");
			}

		}

		protected override void Dispose(bool disposing)
		{
			if ((disposing)) {
				_db.Dispose();
			}
			base.Dispose(disposing);
		}

	}
}
